File: plugin/skills/pr-review/scripts/services/violation_service.py
"""Violation service - transform evaluation results into commentable violations.

Core service - focused on data transformation, no I/O.

Used by:
    - commands/agent/analyze.py (violation creation in pipeline)
"""

from __future__ import annotations

from scripts.commands.agent.comment import CommentableViolation
from scripts.domain.evaluation_task import EvaluationTask
from scripts.services.evaluation_service import EvaluationResult


class ViolationService:
    """Transform evaluation results into commentable violations.

    Core service - focused on data transformation, no I/O.
    All methods are static as they are pure transformations with no state dependency.
    """

    @staticmethod
    def create_violation(
        result: EvaluationResult,
        task: EvaluationTask,
    ) -> CommentableViolation:
        """Create a commentable violation from an evaluation result and task.

        Static - pure transformation, no state needed.

        Args:
            result: The evaluation result containing violation details
            task: The original task containing rule metadata (for documentation_link)

        Returns:
            A CommentableViolation ready for posting to GitHub
        """
        # Extract diff context around the violation line
        diff_context = task.segment.get_context_around_line(
            result.evaluation.line_number,
            context_lines=3,
        )

        return CommentableViolation(
            task_id=result.task_id,
            rule_name=result.rule_name,
            file_path=result.file_path,
            line_number=result.evaluation.line_number,
            score=result.evaluation.score,
            comment=result.evaluation.comment,
            documentation_link=task.rule.documentation_link,
            relevant_claude_skill=task.rule.relevant_claude_skill,
            cost_usd=result.cost_usd,
            diff_context=diff_context,
            rule_url=task.rule.rule_url,
        )

    @staticmethod
    def filter_by_score(
        results: list[EvaluationResult],
        tasks: list[EvaluationTask],
        min_score: int,
    ) -> list[CommentableViolation]:
        """Filter evaluation results and convert to violations.

        Static - pure filter function.

        Args:
            results: List of evaluation results to filter
            tasks: List of tasks (used to look up documentation_link by task_id)
            min_score: Minimum score threshold for inclusion

        Returns:
            List of CommentableViolation objects for results that:
            - Have violates_rule=True
            - Have score >= min_score
        """
        task_map = {task.task_id: task for task in tasks}
        violations: list[CommentableViolation] = []

        for result in results:
            if not result.evaluation.violates_rule:
                continue

            if result.evaluation.score < min_score:
                continue

            task = task_map.get(result.task_id)
            if task:
                violations.append(ViolationService.create_violation(result, task))

        return violations

File: pr-radar-mac/Sources/services/PRRadarCLIService/ViolationService.swift
import Foundation
import PRRadarModels

/// Represents a violation ready for posting as a GitHub comment.
public struct CommentableViolation: Sendable {
    public let taskId: String
    public let ruleName: String
    public let filePath: String
    public let lineNumber: Int?
    public let score: Int
    public let comment: String
    public let documentationLink: String?
    public let relevantClaudeSkill: String?
    public let costUsd: Double?
    public let diffContext: String?
    public let ruleUrl: String?

    public init(
        taskId: String,
        ruleName: String,
        filePath: String,
        lineNumber: Int?,
        score: Int,
        comment: String,
        documentationLink: String? = nil,
        relevantClaudeSkill: String? = nil,
        costUsd: Double? = nil,
        diffContext: String? = nil,
        ruleUrl: String? = nil
    ) {
        self.taskId = taskId
        self.ruleName = ruleName
        self.filePath = filePath
        self.lineNumber = lineNumber
        self.score = score
        self.comment = comment
        self.documentationLink = documentationLink
        self.relevantClaudeSkill = relevantClaudeSkill
        self.costUsd = costUsd
        self.diffContext = diffContext
        self.ruleUrl = ruleUrl
    }

    /// Compose the final GitHub comment body with rule header, comment, and metadata.
    public func composeComment() -> String {
        let ruleHeader: String
        if let ruleUrl {
            ruleHeader = "**[\(ruleName)](\(ruleUrl))**"
        } else {
            ruleHeader = "**\(ruleName)**"
        }

        var lines = [ruleHeader, "", comment]

        if let relevantClaudeSkill {
            lines.append("")
            lines.append("Related Claude Skill: `/\(relevantClaudeSkill)`")
        }

        if let documentationLink {
            lines.append("")
            lines.append("Related Documentation: [Docs](\(documentationLink))")
        }

        let costStr = costUsd.map { String(format: " (cost $%.4f)", $0) } ?? ""
        lines.append("")
        lines.append("*Assisted by [PR Radar](https://github.com/gestrich/PRRadar)\(costStr)*")

        return lines.joined(separator: "\n")
    }
}

/// Pure transformation service for converting evaluation results into commentable violations.
public struct ViolationService: Sendable {
    public init() {}

    /// Create a commentable violation from an evaluation result and task.
    public static func createViolation(
        result: RuleEvaluationResult,
        task: EvaluationTaskOutput
    ) -> CommentableViolation {
        let diffContext = task.focusArea.getContextAroundLine(
            result.evaluation.lineNumber,
            contextLines: 3
        )

        return CommentableViolation(
            taskId: result.taskId,
            ruleName: result.ruleName,
            filePath: result.filePath,
            lineNumber: result.evaluation.lineNumber,
            score: result.evaluation.score,
            comment: result.evaluation.comment,
            documentationLink: task.rule.documentationLink,
            costUsd: result.costUsd,
            diffContext: diffContext
        )
    }

    /// Filter evaluation results by violation status and score, converting to commentable violations.
    public static func filterByScore(
        results: [RuleEvaluationResult],
        tasks: [EvaluationTaskOutput],
        minScore: Int
    ) -> [CommentableViolation] {
        let taskMap = Dictionary(uniqueKeysWithValues: tasks.map { ($0.taskId, $0) })
        var violations: [CommentableViolation] = []

        for result in results {
            guard result.evaluation.violatesRule else { continue }
            guard result.evaluation.score >= minScore else { continue }
            guard let task = taskMap[result.taskId] else { continue }
            violations.append(createViolation(result: result, task: task))
        }

        return violations
    }

    /// Load violations from evaluation result files on disk.
    public static func loadViolations(
        evaluationsDir: String,
        tasksDir: String,
        minScore: Int
    ) -> [CommentableViolation] {
        let fm = FileManager.default
        var violations: [CommentableViolation] = []

        // Load task metadata
        var taskMetadata: [String: EvaluationTaskOutput] = [:]
        if let taskFiles = try? fm.contentsOfDirectory(atPath: tasksDir) {
            for file in taskFiles where file.hasSuffix(".json") {
                let path = "\(tasksDir)/\(file)"
                guard let data = fm.contents(atPath: path),
                      let task = try? JSONDecoder().decode(EvaluationTaskOutput.self, from: data) else { continue }
                taskMetadata[task.taskId] = task
            }
        }

        guard let evalFiles = try? fm.contentsOfDirectory(atPath: evaluationsDir) else { return violations }

        for file in evalFiles where file.hasSuffix(".json") && file != "summary.json" {
            let path = "\(evaluationsDir)/\(file)"
            guard let data = fm.contents(atPath: path),
                  let result = try? JSONDecoder().decode(RuleEvaluationResult.self, from: data) else { continue }

            guard result.evaluation.violatesRule else { continue }
            guard result.evaluation.score >= minScore else { continue }

            let filePath = result.filePath.isEmpty ? result.evaluation.filePath : result.filePath

            let documentationLink: String?
            if let task = taskMetadata[result.taskId] {
                documentationLink = task.rule.documentationLink
            } else {
                documentationLink = nil
            }

            violations.append(CommentableViolation(
                taskId: result.taskId,
                ruleName: result.ruleName,
                filePath: filePath,
                lineNumber: result.evaluation.lineNumber,
                score: result.evaluation.score,
                comment: result.evaluation.comment,
                documentationLink: documentationLink,
                costUsd: result.costUsd
            ))
        }

        return violations
    }
}
